okhttp/src/main/java/okhttp3/Challenge.java

```diff
@@ -16,67 +16,97 @@
 package okhttp3;
 
 import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
 import javax.annotation.Nullable;
 
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.unmodifiableMap;
 import static okhttp3.internal.Util.ISO_8859_1;
 
 /** An RFC 7617 challenge. */
 public final class Challenge {
   private final String scheme;
-  private final String realm;
-  private final Charset charset;
+  private final Map<String, String> authParams;
 
-  public Challenge(String scheme, String realm) {
-    this(scheme, realm, ISO_8859_1);
+  public Challenge(String scheme, Map<String, String> authParams) {
+    if (scheme == null) throw new NullPointerException("scheme == null");
+    this.scheme = scheme.toLowerCase();
+    if (authParams == null) {
```

Comment:
Nullable collections are a bit suspect. Do we want to allow this?

It's also worth noting that the lack of `@Nullable` annotations on this parameters means that the package-level `@NonNull` annotation takes precedence and Java callers and Kotlin callers enforcing JSR 305 _cannot_ pass null even if they wanted to.